syntax = "proto3";

package graplinc.grapl.api.graph_query.v1beta1;

import "graplinc/common/v1beta1/types.proto";
import "graplinc/grapl/api/graph/v1beta1/types.proto";
import "graplinc/grapl/common/v1beta1/types.proto";

message IntegerProperty {
  oneof property {
    graph.v1beta1.IncrementOnlyIntProp increment_only_int = 1;
    graph.v1beta1.DecrementOnlyIntProp decrement_only_int = 2;
    graph.v1beta1.ImmutableIntProp immutable_int = 3;
  }
}

message IntFilter {
  enum Operation {
    UNKNOWN_OPERATION = 0;
    HAS = 1;
    EQUAL = 2;
    LESS_THAN = 3;
    LESS_THAN_OR_EQUAL = 4;
    GREATER_THAN = 5;
    GREATER_THAN_OR_EQUAL = 6;
  }
  Operation operation = 1;
  int64 value = 2;
  bool negated = 3;
}

message AndIntFilters {
  repeated IntFilter int_filters = 1;
}

message OrIntFilters {
  repeated AndIntFilters and_int_filters = 1;
}

message StringFilter {
  enum Operation {
    UNKNOWN_OPERATION = 0;
    HAS = 1;
    EQUAL = 2;
    CONTAINS = 3;
    REGEX = 4;
  }
  Operation operation = 1;
  string value = 2;
  bool negated = 3;
}

message AndStringFilters {
  repeated StringFilter string_filters = 1;
}

message OrStringFilters {
  repeated AndStringFilters and_string_filters = 1;
}

message UidFilter {
  enum Operation {
    UNKNOWN_OPERATION = 0;
    EQUAL = 1;
  }
  Operation operation = 1;
  int64 value = 2;
}

message UidFilters {
  repeated UidFilter uid_filters = 1;
}

message QueryId {
  uint64 value = 1;
}

message NodePropertyQuery {
  graplinc.grapl.common.v1beta1.NodeType node_type = 1;
  uint64 query_id = 2;
  map<string, OrStringFilters> string_filters = 3;
  // todo: Add other property types
}

// Maps to `Dict[Tuple[QueryId, EdgeName], Set[QueryId]]`
message EdgeQueryEntry {
  // Part of the key
  QueryId query_id = 1;
  // Part of the key
  graplinc.grapl.common.v1beta1.EdgeName edge_name = 2;
  // The value
  repeated QueryId neighbor_query_ids = 3;
}

message EdgeQueryMap {
  repeated EdgeQueryEntry entries = 1;
}

message EdgeMapEntry {
  // Key
  graplinc.grapl.common.v1beta1.EdgeName forward_edge_name = 1;
  // Value
  graplinc.grapl.common.v1beta1.EdgeName reverse_edge_name = 2;
}

message EdgeMap {
  repeated EdgeMapEntry edge_map_entries = 1;
}

message GraphQuery {
  QueryId root_query_id = 1;
  map<uint64, NodePropertyQuery> node_property_queries = 2;
  EdgeQueryMap edge_filters = 3;
  EdgeMap edge_map = 4;
}

message NodeView {
  // The uid of the node
  graplinc.grapl.common.v1beta1.Uid uid = 1;
  // The string name of the node's type
  graplinc.grapl.common.v1beta1.NodeType node_type = 2;
  // The string properties of the node
  map<string, string> string_properties = 3;
  // The int properties of the node
  map<string, int64> int_properties = 4;
}

message EdgeView {
  // The name of the edge
  graplinc.grapl.common.v1beta1.EdgeName edge_name = 1;
  // The uid of the source node
  graplinc.grapl.common.v1beta1.Uid source_uid = 2;
  // The uid of the destination node
  graplinc.grapl.common.v1beta1.Uid destination_uid = 3;
}

message EdgeViews {
  repeated EdgeView edge_view_list = 1;
}

message NodeViewEntry {
  graplinc.grapl.common.v1beta1.Uid uid = 1;
  NodeView node_view = 2;
}

message NodeViewMap {
  repeated NodeViewEntry entries = 1;
}

message EdgeViewEntry {
  graplinc.grapl.common.v1beta1.Uid uid = 1;
  graplinc.grapl.common.v1beta1.EdgeName edge_name = 2;
  repeated graplinc.grapl.common.v1beta1.Uid neighbors = 3;
}

message EdgeViewMap {
  repeated EdgeViewEntry entries = 1;
}

message GraphView {
  // The nodes in the graph
  NodeViewMap nodes = 1;
  // The edges in the graph
  EdgeViewMap edges = 2;
}

message QueryGraphWithNodeRequest {
  // The tenant id that the graph is associated with
  graplinc.common.v1beta1.Uuid tenant_id = 1;
  // The node to parameterize against
  graplinc.grapl.common.v1beta1.Uid node_uid = 2;
  // The query to match
  GraphQuery graph_query = 3;
  // Mapping of forward to reverse edges
  map<string, string> edge_mapping = 4;
}

message QueryGraphWithNodeResponse {
  // A view of the MergedGraph that matched our query
  // or None if query did not match
  GraphView matched_graph = 1;
  // The uid of the node that corresponds to the "root" query,
  // or 0 if query did not match
  graplinc.grapl.common.v1beta1.Uid root_uid = 2;
}

message QueryGraphFromNodeRequest {
  // The tenant id that the graph is associated with
  graplinc.common.v1beta1.Uuid tenant_id = 1;
  // The node to parameterize against
  graplinc.grapl.common.v1beta1.Uid node_uid = 2;
  // The query to match
  GraphQuery graph_query = 3;
  // Mapping of forward to reverse edges
  map<string, string> edge_mapping = 4;
}

message QueryGraphFromNodeResponse {
  // A view of the MergedGraph that matched our query
  // or None if query did not match
  GraphView matched_graph = 1;
  // The uid of the node that corresponds to the "root" query,
  // or 0 if query did not match
  graplinc.grapl.common.v1beta1.Uid root_uid = 2;
}

service GraphQueryService {
  // Used to find a node within a graph that matches a query
  rpc QueryGraphWithUid(QueryGraphWithNodeRequest) returns (QueryGraphWithNodeResponse);
  // Performs a query on the node that corresponds to the provided uid as the root
  rpc QueryGraphFromUid(QueryGraphFromNodeRequest) returns (QueryGraphFromNodeResponse);
}
